### ISSUE TITLE
Global variable not accessible from inside the function with @router decorator

### ISSUE THREAD

**Issue by mrvaghani** [tiangolo/fastapi#1040]
### First check

* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.

### Description

How can I import and use a global variable defined inside another python file?
In my testing so far, I am able to access the value of that variable right before the function. But if I access it inside the function decorated with `@router.get`, I get an error for variable not defined.

A [response](https://github.com/tiangolo/fastapi/issues/342#issuecomment-506860645) to another [question about global variables](https://github.com/tiangolo/fastapi/issues/342) say that this should work. What am I doing wrong?
### Additional context

```python
# config.py

import sys
import json

try:
	with open('config.json') as config_file:
		try:
			GLOBAL_CONFIG = json.load(config_file)
		except json.JSONDecodeError:
			print("JSON file does not exist")
except:
	print("error occurred while parsing json:", sys.exc_info()[1])
	GLOBAL_CONFIG = None

```

```python
# items.py

from fastapi import APIRouter
from starlette.responses import JSONResponse
from config import GLOBAL_CONFIG

router = APIRouter()
print(GLOBAL_CONFIG)  # This is valid and has value


@router.get("/", status_code=200)
async def get_items(username):
	# username is a query parameter
	# domain.com/api/items?username=mrvaghani
	if username not in GLOBAL_CONFIG["username"]:
		return JSONResponse(
			status_code=404,
			content={"message": "Not Found"})
```

This causes the variable GLOBAL_CONFIG to be undefined when called inside the function but has the right value for the `print` statement right before the function.

Please let me know if there is any additional information or troubleshooting I can perform to help provide more information on this issue.

**Comment by Dustyposa** [comment]
Try this one:
`app.add_event_handler("startup",  init_config)`
It's work fine for me.

**Comment by mrvaghani** [comment]
> Try this one:
> `app.add_event_handler("startup", init_config)`
> It's work fine for me.

Apologies for the delayed response. Would you happen to know where I can find more information about this command and what other things it could affect?

**Comment by dmontagu** [comment]
@mrvaghani 

As a preface/warning, using global variables like this is generally discouraged for a variety of reasons. You can find many articles discussing why this is the case with a cursory web search, so I won't expand on this here.

-----

Generally, when you can't access an outer variable inside an inner scope in python, it's because you set the variable's somewhere in the inner scope. If you set a variable's value in an inner scope python masks the value from the outer scope. Usually this is the behavior you want, but it can be confusing if you are trying to achieve something different.

To get around this, usually all you need to do is add a `global` statement inside your function:

```python
@router.get("/", status_code=200)
async def get_items(username):
        global GLOBAL_CONFIG  # <-- like this

	# username is a query parameter
	# domain.com/api/items?username=mrvaghani
	if username not in GLOBAL_CONFIG["username"]:
		return JSONResponse(
			status_code=404,
			content={"message": "Not Found"})
```

(Note that in some cases the `nonlocal` keyword may be more appropriate, but you don't need to worry about that for now; `global` is the right choice here.)

It doesn't look like you are doing anything here that would cause the `GLOBAL_CONFIG` to not be found, so maybe something else is going on, but my guess is that you were running into this problem with a slightly different function body?

If not, a traceback might help.

**Comment by mrvaghani** [comment]
> @mrvaghani
> 
> As a preface/warning, using global variables like this is generally discouraged for a variety of reasons. You can find many articles discussing why this is the case with a cursory web search, so I won't expand on this here.
> 
> Generally, when you can't access an outer variable inside an inner scope in python, it's because you assign to the variable somewhere in the inner scope. If you set a variable's value in an inner scope, python masks the variable from the outer scope, preventing you from accidentally overwriting its global value. Usually this is the behavior you want, but it can be confusing if you are trying to achieve something different.
> 
> To get around this, usually all you need to do is add a `global` statement inside your function:
> 
> ```python
> @router.get("/", status_code=200)
> async def get_items(username):
>         global GLOBAL_CONFIG  # <-- like this
> 
> 	# username is a query parameter
> 	# domain.com/api/items?username=mrvaghani
> 	if username not in GLOBAL_CONFIG["username"]:
> 		return JSONResponse(
> 			status_code=404,
> 			content={"message": "Not Found"})
> ```
> 
> (Note that in some cases the `nonlocal` keyword may be more appropriate, but you don't need to worry about that for now; `global` is the right choice here.)
> 
> It doesn't look like you are doing anything here that would cause the `GLOBAL_CONFIG` to not be found, so maybe something else is going on, but my guess is that you were running into this problem with a slightly different function body?
> 
> If not, a traceback might help.

I completely agree with the global config's being a bad idea. I just needed something quick to get my config into the application and this was the only way to deal with it in my environment. I do have plans to move to ENV files at some point in future.

I have modified my original `config.py` and it now seems to be working without having the import inside the function. I will update it with the `global` keyword and do some more testing.

Appreciate the feedback! :)

**Comment by mrvaghani** [comment]
Closing for now as I am no longer able to recreate this issue. Will re-open if needed.

**Comment by tiangolo** [comment]
Thanks for the help here everyone!

Thanks @mrvaghani for reporting back and closing the issue :+1:
